bbs/ssh.cpp

```diff
@@ -43,190 +43,388 @@
 #include "core/log.h"
 #include "core/net.h"
 #include "core/os.h"
+#include "core/strings.h"
+#include "bbs/bbs.h"
+
+#include <libssh/libssh.h>
+#include <libssh/server.h>
 
 using std::string;
 using std::thread;
 using std::unique_ptr;
 
+using namespace wwiv::core;
+using namespace wwiv::strings;
+
 namespace wwiv {
 namespace bbs {
 
 struct socket_error: public std::runtime_error {
   socket_error(const std::string& message): std::runtime_error(message) {}
 };
 
-static constexpr char WWIV_SSH_KEY_NAME[] = "wwiv_ssh_server";
-#define RETURN_IF_ERROR(s) { if (!cryptStatusOK(s)) return false; }
-#define OK(s) cryptStatusOK(s)
+Keys::Keys() {
 
-static bool ssh_once_init() {
-  // This only should be called once.
-  int status = cryptInit();
-  return OK(status);
 }
 
-bool Key::Open() {
-  static bool once = ssh_once_init();
-  CRYPT_KEYSET keyset;
-  int status = cryptKeysetOpen(&keyset, CRYPT_UNUSED, CRYPT_KEYSET_FILE, filename_.c_str(), CRYPT_KEYOPT_NONE);
-  RETURN_IF_ERROR(status);
+// Check that at least one private key exists
+bool Keys::Exist() {
```

Comment:
We could do that with some work. I see the existing codebase as the 'incompatible' base and this being an attempt to fix that. The new codebase (this patch) works with OpenSSH and Putty. As we discussed Syncterm has problems. I'm actually unsure what SSH clients the existing WWIV SSH code works with (just syncterm?)

To address that I have a build of Syncterm against libssh now working with AUTH_NONE and PASSWORD_AUTH, and even PKI! :) . Which is nice because then Syncterm can connect to modern OpenSSH servers too.

Perhaps getting the Syncterm build fixed first, or in parallel mitigates this?